SyntheticTokenAllowanceOwnerMigration fails on testnet

### Description

The `SyntheticTokenAllowanceOwnerMigration` fails on testnet due to a duplicate key violation.

### Steps to reproduce

Run 0.79.0 on testnet data

### Additional context

```bash
2023-05-02T13:56:13.815-0600 ERROR main o.s.b.SpringApplication Application run failed org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'flywayInitializer' defined in class path resource [org/springframework/boot/autoconfigure/flyway/FlywayAutoConfiguration$FlywayConfiguration.class]: Invocation of init method failed; nested exception is org.flywaydb.core.internal.command.DbMigrate$FlywayMigrateException: Migration failed !
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1804)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:620)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:322)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
	at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1156)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:910)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:583)
	at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.refresh(ReactiveWebServerApplicationContext.java:66)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:731)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:408)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:307)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1303)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1292)
	at com.hedera.mirror.importer.MirrorImporterApplication.main(MirrorImporterApplication.java:33)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:108)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:58)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:65)
Caused by: org.flywaydb.core.internal.command.DbMigrate$FlywayMigrateException: Migration failed !
	at org.flywaydb.core.internal.command.DbMigrate.doMigrateGroup(DbMigrate.java:385)
	at org.flywaydb.core.internal.command.DbMigrate.lambda$applyMigrations$1(DbMigrate.java:275)
	at org.flywaydb.core.internal.jdbc.TransactionalExecutionTemplate.execute(TransactionalExecutionTemplate.java:55)
	at org.flywaydb.core.internal.command.DbMigrate.applyMigrations(DbMigrate.java:274)
	at org.flywaydb.core.internal.command.DbMigrate.migrateGroup(DbMigrate.java:247)
	at org.flywaydb.core.internal.command.DbMigrate.lambda$migrateAll$0(DbMigrate.java:141)
	at org.flywaydb.core.internal.database.postgresql.PostgreSQLAdvisoryLockTemplate.execute(PostgreSQLAdvisoryLockTemplate.java:69)
	at org.flywaydb.core.internal.database.postgresql.PostgreSQLConnection.lock(PostgreSQLConnection.java:99)
	at org.flywaydb.core.internal.schemahistory.JdbcTableSchemaHistory.lock(JdbcTableSchemaHistory.java:139)
	at org.flywaydb.core.internal.command.DbMigrate.migrateAll(DbMigrate.java:141)
	at org.flywaydb.core.internal.command.DbMigrate.migrate(DbMigrate.java:98)
	at org.flywaydb.core.Flyway$1.execute(Flyway.java:172)
	at org.flywaydb.core.Flyway$1.execute(Flyway.java:124)
	at org.flywaydb.core.FlywayExecutor.execute(FlywayExecutor.java:205)
	at org.flywaydb.core.Flyway.migrate(Flyway.java:124)
	at org.springframework.boot.autoconfigure.flyway.FlywayMigrationInitializer.afterPropertiesSet(FlywayMigrationInitializer.java:66)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1863)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1800)
	... 26 more
Caused by: org.flywaydb.core.api.FlywayException: Migration failed !
	at org.flywaydb.core.internal.resolver.java.JavaMigrationExecutor.executeOnce(JavaMigrationExecutor.java:69)
	at org.flywaydb.core.internal.resolver.java.JavaMigrationExecutor.lambda$execute$0(JavaMigrationExecutor.java:48)
	at org.flywaydb.core.internal.database.DefaultExecutionStrategy.execute(DefaultExecutionStrategy.java:27)
	at org.flywaydb.core.internal.resolver.java.JavaMigrationExecutor.execute(JavaMigrationExecutor.java:47)
	at org.flywaydb.core.internal.command.DbMigrate.doMigrateGroup(DbMigrate.java:377)
	... 43 more
Caused by: org.springframework.dao.DuplicateKeyException: StatementCallback; SQL [begin;

create temp table token_allowance_temp (
  amount            bigint not null,
  created_timestamp bigint not null,
  owner             bigint not null,
  payer_account_id  bigint not null,
  spender           bigint not null,
  token_id          bigint not null,
  primary key (owner, spender, token_id, created_timestamp)
) on commit drop;

with affected as (
  select ta.*, cr.consensus_timestamp, cr.sender_id
  from (
    select * from token_allowance
    union all
    select * from token_allowance_history
  ) ta
  join contract_result cr on cr.consensus_timestamp = lower(ta.timestamp_range)
), delete_token_allowance as (
  delete from token_allowance ta
  using affected a
  where ta.owner = a.owner and ta.spender = a.spender and ta.token_id = a.token_id
), delete_token_allowance_history as (
  delete from token_allowance_history ta
  using affected a
  where ta.owner = a.owner and ta.spender = a.spender and ta.token_id = a.token_id and ta.timestamp_range = a.timestamp_range
)
insert into token_allowance_temp (amount, created_timestamp, owner, payer_account_id, spender, token_id)
select amount, consensus_timestamp, sender_id, payer_account_id, spender, token_id
from affected;

with correct_timestamp_range as (
  select
    amount,
    owner,
    payer_account_id,
    spender,
    int8range(created_timestamp, (
      select c.created_timestamp
      from token_allowance_temp c
      where c.owner = p.owner and c.spender = p.spender and c.token_id = p.token_id
        and c.created_timestamp > p.created_timestamp
      order by c.created_timestamp
      limit 1)) as timestamp_range,
    token_id
  from token_allowance_temp p
), history as (
  insert into token_allowance_history (amount, owner, payer_account_id, spender, timestamp_range, token_id)
  select * from correct_timestamp_range where upper(timestamp_range) is not null
)
insert into token_allowance (amount, owner, payer_account_id, spender, timestamp_range, token_id)
select * from correct_timestamp_range where upper(timestamp_range) is null;

commit;
]; ERROR: duplicate key value violates unique constraint "token_allowance_pkey"
  Detail: Key (owner, spender, token_id)=(3491438, 3491439, 3491444) already exists.; nested exception is org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "token_allowance_pkey"
  Detail: Key (owner, spender, token_id)=(3491438, 3491439, 3491444) already exists.
	at org.springframework.jdbc.support.SQLErrorCodeSQLExceptionTranslator.doTranslate(SQLErrorCodeSQLExceptionTranslator.java:247)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:70)
	at org.springframework.jdbc.core.JdbcTemplate.translateException(JdbcTemplate.java:1541)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:393)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:431)
	at org.springframework.jdbc.core.JdbcTemplate$$FastClassBySpringCGLIB$$a4771cb8.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy.invokeMethod(CglibAopProxy.java:386)
	at org.springframework.aop.framework.CglibAopProxy.access$000(CglibAopProxy.java:85)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:704)
	at org.springframework.jdbc.core.JdbcTemplate$$EnhancerBySpringCGLIB$$ef0bed9.execute(<generated>)
	at com.hedera.mirror.importer.migration.SyntheticTokenAllowanceOwnerMigration.doMigrate(SyntheticTokenAllowanceOwnerMigration.java:117)
at com.hedera.mirror.importer.migration.MirrorBaseJavaMigration.migrate(MirrorBaseJavaMigration.java:59)
	at com.hedera.mirror.importer.migration.SyntheticTokenAllowanceOwnerMigration.migrate(SyntheticTokenAllowanceOwnerMigration.java:32)
	at org.flywaydb.core.internal.resolver.java.JavaMigrationExecutor.executeOnce(JavaMigrationExecutor.java:55)
	... 47 more
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "token_allowance_pkey"
  Detail: Key (owner, spender, token_id)=(3491438, 3491439, 3491444) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2713)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2401)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:368)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:498)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:415)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:335)
	at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:321)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:297)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:292)
	at com.zaxxer.hikari.pool.ProxyStatement.execute(ProxyStatement.java:94)
	at com.zaxxer.hikari.pool.HikariProxyStatement.execute(HikariProxyStatement.java)
	at org.springframework.jdbc.core.JdbcTemplate$1ExecuteStatementCallback.doInStatement(JdbcTemplate.java:422)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:381)
	... 58 more
```

### Hedera network

testnet

### Version

0.79.0

### Operating system

None

--- a/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/SyntheticTokenAllowanceOwnerMigration.java
+++ b/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/SyntheticTokenAllowanceOwnerMigration.java
@@ -52,15 +52,31 @@ public class SyntheticTokenAllowanceOwnerMigration extends RepeatableMigration {
             ), delete_token_allowance as (
               delete from token_allowance ta
               using affected a
-              where ta.owner = a.owner and ta.spender = a.spender and ta.token_id = a.token_id
+              where ta.owner in (a.owner, a.sender_id) and ta.spender = a.spender and ta.token_id = a.token_id
+              returning
+                ta.amount,
+                lower(ta.timestamp_range) as created_timestamp,
+                a.sender_id as owner,
+                ta.payer_account_id,
+                ta.spender,
+                ta.token_id
             ), delete_token_allowance_history as (
               delete from token_allowance_history ta
               using affected a
-              where ta.owner = a.owner and ta.spender = a.spender and ta.token_id = a.token_id and ta.timestamp_range = a.timestamp_range
+              where (ta.owner = a.owner and ta.spender = a.spender and ta.token_id = a.token_id and ta.timestamp_range = a.timestamp_range) or
+                (ta.owner = a.sender_id and ta.spender = a.spender and ta.token_id = a.token_id)
+              returning
+                ta.amount,
+                lower(ta.timestamp_range) as created_timestamp,
+                a.sender_id as owner,
+                ta.payer_account_id,
+                ta.spender,
+                ta.token_id
             )
             insert into token_allowance_temp (amount, created_timestamp, owner, payer_account_id, spender, token_id)
-            select amount, consensus_timestamp, sender_id, payer_account_id, spender, token_id
-            from affected;
+            select amount, created_timestamp, owner, payer_account_id, spender, token_id from delete_token_allowance
+            union all
+            select amount, created_timestamp, owner, payer_account_id, spender, token_id from delete_token_allowance_history;
 
             with correct_timestamp_range as (
               select

--- a/hedera-mirror-importer/src/main/resources/application.yml
+++ b/hedera-mirror-importer/src/main/resources/application.yml
@@ -24,6 +24,8 @@ hedera:
           checksum: 3
         initializeEntityBalanceMigration:
           checksum: 3
+        syntheticTokenAllowanceOwnerMigration:
+          checksum: 2
         tokenAccountBalanceMigration:
           checksum: 3
 

--- a/hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/migration/SyntheticTokenAllowanceOwnerMigrationTest.java
+++ b/hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/migration/SyntheticTokenAllowanceOwnerMigrationTest.java
@@ -42,7 +42,7 @@ class SyntheticTokenAllowanceOwnerMigrationTest extends IntegrationTest {
 
     @Test
     void checksum() {
-        assertThat(migration.getChecksum()).isEqualTo(1);
+        assertThat(migration.getChecksum()).isEqualTo(2);
     }
 
     @Test
@@ -78,6 +78,41 @@ class SyntheticTokenAllowanceOwnerMigrationTest extends IntegrationTest {
                         .consensusTimestamp(newTokenAllowance.getTimestampLower()))
                 .persist();
 
+        // A token allowance that has no contract result, but has primary key values that will end up matching those of
+        // a migrated token allowance.
+        var ownerAccountId = 38L;
+        var spender = 39L;
+        var tokenId = 44L;
+        var tokenAllowancePreMigration = domainBuilder
+                .tokenAllowance()
+                .customize(t -> t.owner(ownerAccountId)
+                        .payerAccountId(EntityId.of(ownerAccountId, CONTRACT))
+                        .spender(spender)
+                        .timestampRange(Range.atLeast(1676546171829734003L))
+                        .tokenId(tokenId))
+                .persist();
+
+        // A token allowance that has a contract result, once migrated it will have the same primary key fields as the
+        // above token allowance.
+        var ownerPreMigration = 1322L;
+        var contractResultConsensus = 1676546391434923004L;
+        var collidedTokenAllowance = domainBuilder
+                .tokenAllowance()
+                .customize(t -> t.owner(ownerPreMigration)
+                        .payerAccountId(EntityId.of(ownerPreMigration, CONTRACT))
+                        .spender(spender)
+                        .timestampRange(Range.atLeast(contractResultConsensus))
+                        .tokenId(tokenId))
+                .persist();
+
+        // The contract result for the collided token allowance
+        domainBuilder
+                .contractResult()
+                .customize(c -> c.senderId(EntityId.of(ownerAccountId, CONTRACT))
+                        .payerAccountId(EntityId.of(ownerPreMigration, CONTRACT))
+                        .consensusTimestamp(contractResultConsensus))
+                .persist();
+
         // when
         migration.doMigrate();
 
@@ -88,6 +123,8 @@ class SyntheticTokenAllowanceOwnerMigrationTest extends IntegrationTest {
         expected.add(unaffectedTokenAllowancePair.getLeft());
         expected.add(correctTokenAllowancePair.getLeft());
         expected.add(newTokenAllowance);
+        collidedTokenAllowance.setOwner(ownerAccountId);
+        expected.add(collidedTokenAllowance);
         assertThat(tokenAllowanceRepository.findAll()).containsExactlyInAnyOrderElementsOf(expected);
 
         // The history of the token allowance should also be updated with the corrected owner
@@ -96,6 +133,8 @@ class SyntheticTokenAllowanceOwnerMigrationTest extends IntegrationTest {
         expectedHistory.forEach(t -> t.setOwner(contractResultSenderId.getId()));
         expectedHistory.addAll(unaffectedTokenAllowancePair.getRight());
         expectedHistory.addAll(correctTokenAllowancePair.getRight());
+        tokenAllowancePreMigration.setTimestampUpper(contractResultConsensus);
+        expectedHistory.add(tokenAllowancePreMigration);
         assertThat(findHistory(TokenAllowance.class, idColumns)).containsExactlyInAnyOrderElementsOf(expectedHistory);
     }
 
@@ -113,6 +152,37 @@ class SyntheticTokenAllowanceOwnerMigrationTest extends IntegrationTest {
                 .contractResult()
                 .customize(c -> c.senderId(new EntityId(0L, 0L, newTokenAllowance.getOwner(), CONTRACT))
                         .consensusTimestamp(newTokenAllowance.getTimestampLower()))
+                .persist();
+
+        // The collision token allowances
+        var ownerAccountId = 3491438L;
+        var spender = 3491439L;
+        var tokenId = 3491444L;
+        domainBuilder
+                .tokenAllowance()
+                .customize(t -> t.owner(ownerAccountId)
+                        .payerAccountId(EntityId.of(ownerAccountId, CONTRACT))
+                        .spender(spender)
+                        .timestampRange(Range.atLeast(1676546171829734003L))
+                        .tokenId(tokenId))
+                .persist();
+
+        var ownerPreMigration = 1322L;
+        var contractResultConsensus = 1676546391434923004L;
+        domainBuilder
+                .tokenAllowance()
+                .customize(t -> t.owner(ownerPreMigration)
+                        .payerAccountId(EntityId.of(ownerPreMigration, CONTRACT))
+                        .spender(spender)
+                        .timestampRange(Range.atLeast(contractResultConsensus))
+                        .tokenId(tokenId))
+                .persist();
+
+        domainBuilder
+                .contractResult()
+                .customize(c -> c.senderId(EntityId.of(ownerAccountId, CONTRACT))
+                        .payerAccountId(EntityId.of(ownerPreMigration, CONTRACT))
+                        .consensusTimestamp(contractResultConsensus))
                 .persist();
 
         migration.doMigrate();

